spark-kernel-dsv2/src/test/java/io/delta/spark/dsv2/utils/ExpressionUtilsTest.java

```diff
@@ -0,0 +1,213 @@
+        Arguments.of(
+            "GreaterThanOrEqual", (Supplier<Filter>) () -> new GreaterThanOrEqual("age", 18), ">="),
+        Arguments.of("LessThan", (Supplier<Filter>) () -> new LessThan("age", 65), "<"),
+        Arguments.of(
+            "LessThanOrEqual", (Supplier<Filter>) () -> new LessThanOrEqual("age", 65), "<="));
+  }
+
+  @ParameterizedTest(name = "{0} filter should be converted to {2}")
+  @MethodSource("comparisonFiltersProvider")
+  public void testComparisonFilters(
+      String filterName, Supplier<Filter> filterSupplier, String expectedOperator) {
+    Filter filter = filterSupplier.get();
+    Optional<Predicate> result = ExpressionUtils.convertSparkFilterToKernelPredicate(filter);
+
+    assertTrue(result.isPresent(), filterName + " filter should be converted");
+    assertEquals(expectedOperator, result.get().getName());
+    assertEquals(2, result.get().getChildren().size());
+  }
+
+  // Test data provider for null filters
+  static Stream<Arguments> nullFiltersProvider() {
+    return Stream.of(
+        Arguments.of("IsNull", (Supplier<Filter>) () -> new IsNull("name"), "IS_NULL"),
+        Arguments.of("IsNotNull", (Supplier<Filter>) () -> new IsNotNull("name"), "IS_NOT_NULL"));
+  }
+
+  @ParameterizedTest(name = "{0} filter should be converted to {2}")
+  @MethodSource("nullFiltersProvider")
+  public void testNullFilters(
+      String filterName, Supplier<Filter> filterSupplier, String expectedOperator) {
+    Filter filter = filterSupplier.get();
+    Optional<Predicate> result = ExpressionUtils.convertSparkFilterToKernelPredicate(filter);
+
+    assertTrue(result.isPresent(), filterName + " filter should be converted");
+    assertEquals(expectedOperator, result.get().getName());
+    assertEquals(1, result.get().getChildren().size());
+  }
+
+  @Test
+  public void testEqualNullSafeFilter() {
+    // Test EqualNullSafe with null value - converted to IS_NULL
+    // Cannot use IS NOT DISTINCT FROM because kernel requires typed null literals
+    EqualNullSafe nullFilter = new EqualNullSafe("name", null);
+    Optional<Predicate> nullResult =
+        ExpressionUtils.convertSparkFilterToKernelPredicate(nullFilter);
+
+    assertTrue(nullResult.isPresent(), "EqualNullSafe with null should be converted");
+    assertEquals("IS_NULL", nullResult.get().getName());
+    assertEquals(1, nullResult.get().getChildren().size());
+
+    // Test EqualNullSafe with non-null value - uses IS NOT DISTINCT FROM
+    EqualNullSafe nonNullFilter = new EqualNullSafe("id", 42);
+    Optional<Predicate> nonNullResult =
+        ExpressionUtils.convertSparkFilterToKernelPredicate(nonNullFilter);
+
+    assertTrue(nonNullResult.isPresent(), "EqualNullSafe with value should be converted");
+    assertEquals("=", nonNullResult.get().getName());
+    assertEquals(2, nonNullResult.get().getChildren().size());
+  }
+
+  @Test
+  public void testAndFilter() {
```

Comment:
QQ: shall we add test case that a filter can't be converted to kernel predicate? Or can only be partially converted to kernel predicate?